File: ruby/difference-of-squares/difference_of_squares.rb
module BookKeeping
  VERSION = 4
end

class Squares
  def initialize(num)
    @num = num
  end

  def square_of_sum
    (0..@num).sum ** 2
  end

  def sum_of_squares
    (0..@num).sum {|item| item ** 2}
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: c/difference-of-squares/src/difference_of_squares.c
#include "difference_of_squares.h"

unsigned long long square_of_sum(unsigned int n) {
  unsigned long long total = (n * (n + 1)) / 2;
  return total * total;
}

unsigned long long sum_of_squares(unsigned int n) {
  unsigned long long total = (n * (n + 1) * (2 * n + 1)) / 6;
  return total;
}

unsigned long long difference_of_squares(unsigned int n) {
  return square_of_sum(n) - sum_of_squares(n);
}
